File: python/systemdesign/distributed_lock_simulation.py
"""
Simulation of a distributed lock service using lease-based locks with
expiry, similar in spirit to a Redis/Zookeeper-backed mutex. A client
can only acquire a lock if it's free or its lease has expired, and
only the current lease holder can release or renew it, preventing two
clients from holding the same lock at once.

status - completed
"""

from base_logger.logging_event import create_logger


class LockDeniedError(Exception):
    pass


class DistributedLockService:

    log = create_logger(__name__)

    def __init__(self, clock):
        self.clock = clock
        self.locks = {}

    def _is_expired(self, resource):
        entry = self.locks.get(resource)
        return entry is not None and self.clock() >= entry["expires_at"]

    def acquire(self, resource, client_id, lease_seconds):
        entry = self.locks.get(resource)
        if entry is not None and not self._is_expired(resource) and entry["client_id"] != client_id:
            raise LockDeniedError(
                f"resource '{resource}' held by another client until {entry['expires_at']}"
            )

        self.locks[resource] = {
            "client_id": client_id,
            "expires_at": self.clock() + lease_seconds,
        }
        return True

    def release(self, resource, client_id):
        entry = self.locks.get(resource)
        if entry is None:
            return False
        if entry["client_id"] != client_id and not self._is_expired(resource):
            raise LockDeniedError(f"client {client_id} does not hold the lock on '{resource}'")
        del self.locks[resource]
        return True

    def is_locked(self, resource):
        return resource in self.locks and not self._is_expired(resource)


if __name__ == "__main__":
    log = create_logger("distributed_lock_demo")

    fake_time = [0.0]

    def fake_clock():
        return fake_time[0]

    lock_service = DistributedLockService(clock=fake_clock)

    lock_service.acquire("resource-1", client_id="client-A", lease_seconds=5)
    log.info("client-A acquired resource-1, is_locked=%s", lock_service.is_locked("resource-1"))

    try:
        lock_service.acquire("resource-1", client_id="client-B", lease_seconds=5)
    except LockDeniedError as exc:
        log.info("client-B denied while lease active: %s", exc)

    fake_time[0] = 6.0
    log.info("after lease expires (t=6), client-B acquires: %s",
             lock_service.acquire("resource-1", client_id="client-B", lease_seconds=5))

    try:
        lock_service.release("resource-1", client_id="client-A")
    except LockDeniedError as exc:
        log.info("client-A cannot release a lock it no longer holds: %s", exc)

    released = lock_service.release("resource-1", client_id="client-B")
    log.info("client-B releases resource-1: %s, is_locked=%s",
             released, lock_service.is_locked("resource-1"))

File: typescript/systemdesign/distributedLockSimulation.ts
/*
Simulates a distributed lock service with lease/expiry semantics: a client
acquires a named lock with a time-limited lease (fencing token included),
must renew before it expires to keep holding it, and an expired lease can
be stolen by another client - preventing two clients from holding the same
lock at once.

status - completed
*/

import { logp } from "../utils/logger";

interface Lease {
  ownerId: string;
  fencingToken: number;
  expiresAtMs: number;
}

class LockAcquisitionError extends Error {}

class DistributedLockSimulation {
  private readonly leasesByLockName: Map<string, Lease> = new Map();
  private nextFencingToken = 1;

  constructor(private readonly nowFn: () => number = Date.now) {}

  private isExpired(lease: Lease): boolean {
    return this.nowFn() >= lease.expiresAtMs;
  }

  acquire(lockName: string, clientId: string, leaseDurationMs: number): number {
    const existing = this.leasesByLockName.get(lockName);
    if (existing && !this.isExpired(existing) && existing.ownerId !== clientId) {
      throw new LockAcquisitionError(`lock "${lockName}" is held by "${existing.ownerId}"`);
    }

    const fencingToken = this.nextFencingToken++;
    this.leasesByLockName.set(lockName, {
      ownerId: clientId,
      fencingToken,
      expiresAtMs: this.nowFn() + leaseDurationMs,
    });
    return fencingToken;
  }

  renew(lockName: string, clientId: string, leaseDurationMs: number): void {
    const existing = this.leasesByLockName.get(lockName);
    if (!existing || this.isExpired(existing) || existing.ownerId !== clientId) {
      throw new LockAcquisitionError(`"${clientId}" does not hold a valid lease on "${lockName}"`);
    }
    existing.expiresAtMs = this.nowFn() + leaseDurationMs;
  }

  release(lockName: string, clientId: string): void {
    const existing = this.leasesByLockName.get(lockName);
    if (existing && existing.ownerId === clientId) {
      this.leasesByLockName.delete(lockName);
    }
  }

  currentOwner(lockName: string): string | undefined {
    const existing = this.leasesByLockName.get(lockName);
    if (!existing || this.isExpired(existing)) {
      return undefined;
    }
    return existing.ownerId;
  }
}

// usage scenarios: fake clock, two competing clients

let simulatedNow = 0;
const lockService = new DistributedLockSimulation(() => simulatedNow);

const tokenA = lockService.acquire("job-runner", "client-A", 500);
logp(`client-A acquired lock with fencing token ${tokenA}`);

try {
  lockService.acquire("job-runner", "client-B", 500);
} catch (err) {
  logp(`client-B failed to acquire while A holds it: ${(err as Error).message}`);
}

logp(`current owner -> ${lockService.currentOwner("job-runner")}`);

logp("advancing clock by 200ms, client-A renews before expiry");
simulatedNow += 200;
lockService.renew("job-runner", "client-A", 500);
logp(`current owner still -> ${lockService.currentOwner("job-runner")}`);

logp("advancing clock by 600ms without renewal, lease expires");
simulatedNow += 600;
logp(`current owner after expiry -> ${lockService.currentOwner("job-runner")} (expected undefined)`);

const tokenB = lockService.acquire("job-runner", "client-B", 500);
logp(`client-B acquired expired lock, new fencing token ${tokenB} (must be greater than A's token ${tokenA}: ${tokenB > tokenA})`);

try {
  lockService.renew("job-runner", "client-A", 500);
} catch (err) {
  logp(`client-A can no longer renew, lock stolen: ${(err as Error).message}`);
}
